habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py

```diff
@@ -241,13 +274,169 @@ def add_receptacle_visualization(
             )
         return [box_obj]
 
+    def debug_draw(
+        self, sim: habitat_sim.Simulator, color: Optional[mn.Color4] = None
+    ) -> None:
+        """
+        Render the AABBReceptacle with DebugLineRender utility at the current frame.
+        Must be called after each frame is rendered, before querying the image data.
+
+        :param sim: Simulator must be provided.
+        :param color: Optionally provide wireframe color, otherwise magenta.
+        """
+        # draw the box
+        if color is None:
+            color = mn.Color4.magenta()
+        dblr = sim.get_debug_line_render()
+        dblr.push_transform(self.get_global_transform(sim))
+        dblr.draw_box(self.bounds.min, self.bounds.max, color)
+        dblr.pop_transform()
+        # TODO: test this
+
 
-def get_all_scenedataset_receptacles(sim) -> Dict[str, Dict[str, List[str]]]:
+class TriangleMeshReceptacle(Receptacle):
+    """
+    Defines a Receptacle surface as a triangle mesh.
+    TODO: configurable maximum height.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        mesh_data: Tuple[List[Any], List[Any]],  # vertices, indices
+        parent_object_handle: str = None,
+        parent_link: Optional[int] = None,
+        up: Optional[mn.Vector3] = None,
+    ) -> None:
+        """
+        Initialize the TriangleMeshReceptacle from mesh data and pre-compute the area weighted accumulator.
+
+        :param name: The name of the Receptacle. Should be unique and descriptive for any one object.
+        :param mesh_data: The Receptacle's mesh data. A Tuple of two Lists, first vertex geometry (Vector3) and second topology (indicies of triangle corner verts(int) (len divisible by 3)).
+        :param parent_object_handle: The rigid or articulated object instance handle for the parent object to which the Receptacle is attached. None for globally defined stage Receptacles.
+        :param parent_link: Index of the link to which the Receptacle is attached if the parent is an ArticulatedObject. -1 denotes the base link. None for rigid objects and stage Receptables.
+        :param up: The "up" direction of the Receptacle in local AABB space. Used for optionally culling receptacles in un-supportive states such as inverted surfaces.
+        """
+        super().__init__(name, parent_object_handle, parent_link, up)
+        self.mesh_data = mesh_data
+        self.area_weighted_accumulator = (
+            []
+        )  # normalized float weights for each triangle for sampling
+        assert len(mesh_data[1]) % 3 == 0, "must be triangles"
```

Comment:
```suggestion
        assert len(mesh_data[1]) % 3 == 0, "Provided mesh is not a Triangle. Triangle Mesh receptacles can only be triangles"
```

Reply:
I'm thinking that this should ideally be something to handle directly in the Magnum asset pipeline, so you can always be sure that a `MeshPrimitive.TRIANGLES` mesh has the index / vertex count divisible by 3.

Currently that is done only partially (it's always true for PLY, STL or OBJ but for example not for glTF) and it's not enforced by MeshData itself, but I could clean that up. Seems like desirable behavior to me.